### ISSUE TITLE
Improve computing efficiency using `torch.infinite`

### ISSUE THREAD

**Issue by namgyu-youn** [facebookresearch/optimizers#99]
In `utils/shampoo_preconditioner_list.py [Ln 687]`, checking for `ErrorType(NaN & Inf)` worked like a **multiple boolean process**. It can be written as :
```
if torch.isnan(factor_matrix).any():
... (raise)
if torch.isinf(factor_matrix).any():
... (raise)
```

But in my experience, [torch.infinite](https://pytorch.org/docs/stable/generated/torch.isfinite.html) was really useful, like this (`NaN & Inf`) case. So I think this method could be improved. Also, `torch.isinfinite` is useful when most of the test cases are correct. It works only when the `ErrorType` happens.

**Comment by namgyu-youn** [comment]
> I wonder what do you mean improving computing efficiency?

The point is, Shampoo should be stopped when `NaN or Inf` is detected. So it is implemented using `elif` now. 

That means if `NaN` is detected before, `Inf` doesn't have to be progressed because an error was already detected.

As sure, this could be fixed using simple `error_check`, but I think `torch.infinite` could be better because it helps more clean code.

**Comment by namgyu-youn** [comment]
> I wonder what do you mean improving computing efficiency?

The point is, the current error detection process is based on multi-step (`NaN -> Inf`). So `Inf` detection works until the step is finished.

If `NaN & Inf` detection could be done in a single step, it could be more efficient based on Tensor's traversal.

**Comment by namgyu-youn** [comment]
> Because if you actually think about how to implement `torch.isfinite` by yourself, the implementation might decompose into two parts of checking input is an `inf` and checking the input is an `nan`, and from computational complexity-perspective, this is not significant better than we have now.

I think this issue should address the CUDA-level approach. In CUDA, `Inf & NaN` has **specific patterns**.
 
In IEEE 754, `Inf (Infinity)` and `NaN (Not a Number)` have specific patterns. They can be written in bit-pattern : 
- Positive Infinity : `0 11111111 00000000000000000000000`
- Negative Infinity : `1 11111111 00000000000000000000000`
- NaN (Not a Number) : `0 11111111 10000000000000000000000`

As you can see above, many bits are the same. So they could be computed at the same time in CUDA. `torch.isfinite` focus on this, and is built for more efficient computation in this case.

**Comment by namgyu-youn** [comment]
> The `do something` breaks the branch predication mechanism, and what `torch.isfinite` is technically fusing the checks of `torch.isnan(factor_matrix).any()` and `torch.isinf(factor_matrix).any()` together before user does `do something`. As a result, this must be not slower compared to what we had currently. It just in the macro-Shampoo level perspective, this part is and will not likely be the bottlenecks in terms of performance; however, personally I really like the brevity because for Shampoo, encountering either `nan` or `inf` is a hard-stop, that is the reason we raise.

If I understood it right, it could be summarized like the following:
- Error (NaN & Inf) is a sparse case. Therefore, most of the Shampoo passes the full detection process (`isnan & isinf`).
- Because the error is a sparse case, `(do something)` rarely happens. (e.g., `isnan -> error -> raise (no process for isinf)`)
- Regardless of sparsity, `(do something)` means independent process only when an error occurs.

Then, what if `isnan & isinf` could be processed simultaneously at almost the same speed as two-step? I am not saying bottleneck. Because PyTorch is wrapped framework of CUDA, CUDA can transform current two-step just single step.

**Comment by namgyu-youn** [comment]
> Yes, it should be pretty rare from Shampoo perspective; however, in practice, one common case due to user-side errors is bad data (e.g., out-of-distribution data) which likely causes close-to-infinity gradient then those data will trigger this.

> I will say CUDA **can** processed those checks simultaneously, and the original pattern **definitely** is not able to achieve this due to user-side logic (`do something` in this context no matter how unlikely `isnan` and `isinf`).

I see clear, thanks. Maybe the goal is **offering specific ErrorType**, not computation efficiency. The user must get a specific ErrorType whether the detection process is simultaneous or not, right? 

Therefore, if a user can get a specific ErrorType and code could be brevity, this update could be useful. Is this true?

**Comment by namgyu-youn** [comment]
> If our goal is offering specific ErrorType, then the current pattern (i.e., treating `inf` and `nan` separately) is better because `inf` and `nan` are different and users might have their own domain knowledges and experience on **why** encountering `inf` and `nan` separately so separating those two gives users a better chance to debug when encountering `inf` and `nan`. If we check `_check_factor_matrix_for_diagonality_nan_and_inf`, you should see the error messages are slightly different between encounter `inf` and `nan` because we injected our domain knowledge and experience on dealing with those errors.
> We do also have specific ErrorType, raised from the critical error part that we don't want general catching on `ValueError` interrupt this. We want users to explicitly think about what they want to do when encountering `PreconditionerValueError`.

Thanks for your explanation. I understand this works already for offering specific ErrorType. But I thought error message could be updated using binary method in `infinite` case, because only some markdowns (not `factor_matrix`) are slightly different.

And the main point is computational efficiency.  For more explaination:
```
# Current Version - Multi step
if torch.isnan().any(): # step
   raise ...
if torch.isinf().any(): # step
   raise ...

# Updated Version
if not torch.isfinite().all():  # step
    has_nan = torch.isnan().any()  # It works only when error is detected
    error_type = "nan" if has_nan else "inf"
    raise ...
```

Because the error is a **sparse** case, **most of Shampoo would pass single step (if not isfinite) in the updated version.** 

Of course, an additional process for configuring error type (`NaN or Inf`) is needed, but it is required only when an error occurs.

**Comment by namgyu-youn** [comment]
> if the fused function (`torch.isfinite` in this case) is way more efficient and this is the performance bottleneck, we should consider using the updated version; however, this is not the case right now, and personally I prefer not sacrificing the code clarity for this insignificant performance gain.

I agreed with that. Then I would close the issue. If you think this should be handled, then open and ping me please.

**Comment by namgyu-youn** [comment]
Also, the fixed version would be :
```
[Ln 687 ~ 700]
        if not torch.isfinite(factor_matrix).all():
            has_nan = torch.isnan(factor_matrix).any()
            error_type = "nan" if has_nan else "inf"

            mitigation_msg = (
                "To mitigate, check if nan inputs are being passed into the network or nan gradients "
                "are being passed to the optimizer."
                if has_nan
                else "In some cases, this may be due to divergence of the algorithm. "
                "To mitigate, try decreasing the learning rate or increasing grafting epsilon."
            )

            raise PreconditionerValueError(
                f"Encountered {error_type} values in factor matrix {factor_matrix_index}! "
                f"{mitigation_msg} "

**Comment by namgyu-youn** [comment]
> Sorry I did not make it clear before, I think the brevity is still beneficial for us to do it, and I am considering let's just merging the error messages of both `nan` and `inf` together without further differentiating `nan` and `inf` if `torch.isfinite() == True`. What I am thinking is as follows:

Never mind. I should have attached it first for explanation. And I am considering exactly the same with you.

> Let me know if you want to refactor this, and I think there might be few tests we have to fix to make it work.

Yes. This update should be done within accurate testing for NaN, Inf, and both.

But I am sorry you opened a PR for this without even co-auther. I attached to you the full explanation and also the source code. Please care my invested time and effort. I want co contribute more, and this makes me hestiated.
